src/legacy/core_plugins/data/public/search/aggs/buckets/lib/time_buckets/time_buckets.ts

```diff
@@ -0,0 +1,460 @@
+/*
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import _ from 'lodash';
+import moment from 'moment';
+
+import { IUiSettingsClient } from '../../../../../../../../../core/public';
+import {
+  fieldFormats as fieldFormatsHelpers,
+  parseInterval,
+} from '../../../../../../../../../plugins/data/public';
+// eslint-disable-next-line @kbn/eslint/no-restricted-paths
+import { FieldFormatsStart } from '../../../../../../../../../plugins/data/public/field_formats';
+import { calcAutoIntervalLessThan, calcAutoIntervalNear } from './calc_auto_interval';
+import {
+  convertDurationToNormalizedEsInterval,
+  convertIntervalToEsInterval,
+  EsInterval,
+} from './calc_es_interval';
+
+interface Bounds {
+  min: Date | number | null;
+  max: Date | number | null;
+}
+
+interface TimeBucketsInterval extends moment.Duration {
+  // TODO double-check whether all of these are needed
+  description: string;
+  esValue: EsInterval['value'];
+  esUnit: EsInterval['unit'];
+  expression: EsInterval['expression'];
+  overflow: moment.Duration | boolean;
+  preScaled?: moment.Duration;
+  scale?: number;
+  scaled?: boolean;
+}
+
+function isObject(o: any): o is Record<string, any> {
+  return _.isObject(o);
+}
+
+function isString(s: any): s is string {
+  return _.isString(s);
+}
+
+function isValidMoment(m: any): boolean {
+  return m && 'isValid' in m && m.isValid();
+}
+
+interface TimeBucketsConfig {
+  fieldFormats: FieldFormatsStart;
+  uiSettings: IUiSettingsClient;
+}
+
+/**
+ * Helper class for wrapping the concept of an "Interval",
+ * which describes a timespan that will separate moments.
+ *
+ * @param {state} object - one of ""
+ * @param {[type]} display [description]
+ */
+export class TimeBuckets {
+  private getConfig: (key: string) => any;
+  private fieldFormats: FieldFormatsStart;
+
+  private _lb: Bounds['min'] = null;
+  private _ub: Bounds['max'] = null;
+  private _originalInterval: string | null = null;
+  private _i?: moment.Duration | 'auto';
+
+  // because other parts of Kibana arbitrarily add properties
+  [key: string]: any;
+
+  static __cached__(self: TimeBuckets) {
```

Comment:
i think we can get rid of the __cached__ part

Reply:
as discussed offline, we'll do this as a follow up step